runtime/protocol/http/common/test/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecutionTest.kt

```diff
@@ -0,0 +1,103 @@
+class SdkOperationExecutionTest {
+
+    val retryableTestException = ServiceException("test exception").apply {
+        sdkErrorMetadata.attributes[ErrorMetadata.Retryable] = true
+        sdkErrorMetadata.attributes[ServiceErrorMetadata.ErrorType] = ServiceException.ErrorType.Server
+    }
+
+    @Test
+    fun testOperationMiddlewareOrder() = runTest {
+        // sanity test middleware flows the way we expect
+        val serialized = HttpRequestBuilder()
+        val op = newTestOperation<Unit, Unit>(serialized, Unit)
+
+        val mockEngine = object : HttpClientEngineBase("test engine") {
+            private var attempt = 0
+            override suspend fun roundTrip(context: ExecutionContext, request: HttpRequest): HttpCall {
+                attempt++
+                if (attempt == 1) throw retryableTestException
+                val resp = HttpResponse(HttpStatusCode.OK, Headers.Empty, HttpBody.Empty)
+                return HttpCall(request, resp, Instant.now(), Instant.now())
+            }
+        }
+
+        op.execution.signer = object : HttpSigner {
+            override suspend fun sign(context: ExecutionContext, request: HttpRequestBuilder) {
+                assertFalse(request.headers.contains("test-auth"))
+                request.headers.append("test-auth", "test-signature")
+
+                assertFalse(request.headers.contains("receive-header"))
+            }
+        }
+
+        val actualOrder = mutableListOf<String>()
+
+        op.execution.initialize.intercept { req, next ->
+            actualOrder.add("initialize")
+            next.call(req)
+        }
+
+        op.execution.mutate.intercept { req, next ->
+            actualOrder.add("mutate")
+            req.subject.headers.append("mutate-header", "mutate")
+            next.call(req)
+        }
+
+        op.execution.onEachAttempt.intercept { req, next ->
+            actualOrder.add("attempt")
+            // retry middleware should be giving us a fresh copy each attempt
+            assertFalse(req.subject.headers.contains("attempt-header"))
+            req.subject.headers.append("attempt-header", "per-attempt")
+
+            // signing should come after this
+            assertFalse(req.subject.headers.contains("test-auth"))
+            next.call(req)
+        }
+
+        op.execution.receive.intercept { req, next ->
+            actualOrder.add("receive")
+            assertTrue(req.subject.headers.contains("attempt-header"))
+            assertTrue(req.subject.headers.contains("test-auth"))
+            assertTrue(req.subject.headers.contains("mutate-header"))
+
+            assertFalse(req.subject.headers.contains("receive-header"))
+            req.subject.headers.append("receive-header", "receive")
+            next.call(req)
+        }
+
+        val client = sdkHttpClient(mockEngine)
+        op.roundTrip(client, Unit)
+        val expectedOrder = listOf("initialize", "mutate", "attempt", "receive", "attempt", "receive")
```

Comment:
**Question**: Is it worth adding verification of the reverse-order call as well?

Reply:
We could. I thought about it and then decided not to since it's a call stack...